lxd/api_cluster.go

```diff
@@ -4365,3 +4366,102 @@ func clusterSelectTarget(ctx context.Context, s *state.State, gateway *cluster.G
 
 	return targetMemberInfo, nil
 }
+
+func autoHealClusterTask(d *Daemon) (task.Func, task.Schedule) {
+	f := func(ctx context.Context) {
+		leader, err := d.gateway.LeaderAddress()
+		if err != nil {
+			if errors.Is(err, cluster.ErrNodeIsNotClustered) {
+				return // No error if not clustered.
+			}
+
+			logger.Error("Failed to get leader cluster member address", logger.Ctx{"err": err})
+			return
+		}
+
+		if d.State().LocalConfig.ClusterAddress() != leader {
+			logger.Debug("Skipping heal cluster task since we're not leader")
+			return
+		}
+
+		opRun := func(op *operations.Operation) error {
+			return autoHealCluster(ctx, d)
+		}
+
+		op, err := operations.OperationCreate(d.State(), "", operations.OperationClassTask, operationtype.ClusterHeal, nil, nil, opRun, nil, nil, nil)
+		if err != nil {
+			logger.Error("Failed to start heal cluster operation", logger.Ctx{"err": err})
+			return
+		}
+
+		err = op.Start()
+		if err != nil {
+			logger.Error("Failed to heal cluster", logger.Ctx{"err": err})
+			return
+		}
+
+		_, _ = op.Wait(ctx)
+	}
+
+	return f, task.Every(time.Minute)
+}
+
+func autoHealCluster(ctx context.Context, d *Daemon) error {
+	s := d.State()
+	offlineThreshold := s.GlobalConfig.OfflineThreshold()
+	healingThreshold := s.GlobalConfig.ClusterHealing()
+
+	// Skip healing if it's disabled.
+	if healingThreshold == 0 {
+		return nil
+	}
+
+	var members []db.NodeInfo
+	var err error
+
+	err = s.DB.Cluster.Transaction(context.TODO(), func(ctx context.Context, tx *db.ClusterTx) error {
+		members, err = tx.GetNodes(ctx)
+		if err != nil {
+			return fmt.Errorf("Failed getting cluster members: %w", err)
+		}
+
+		return nil
+	})
+	if err != nil {
+		return fmt.Errorf("Failed healing cluster: %w", err)
+	}
+
+	var offlineMembers []db.NodeInfo
+
+	for _, member := range members {
+		// Ignore online members, those which have been evacuated, and those haven't exceeded the healing threshold.
+		if !member.IsOffline(offlineThreshold) || member.State == db.ClusterMemberStateEvacuated || !member.IsOffline(healingThreshold) {
```

Comment:
We only need to check `!member.IsOffline(healingThreshold)` here as the `s.GlobalConfig.ClusterHealing()` will always return the greater of s.GlobalConfig.OfflineThreshold() or cluster healing duration. No need to to also check `!member.IsOffline(offlineThreshold)`